storage: use a "distinct" batch in Replica.handleRaftReady()

petermattis opened the pull request:
A minor optimization that affects small transactions.

Fixes #6949.

```
name                       old time/op    new time/op    delta
Insert1_Cockroach-8           212µs ± 2%     204µs ± 1%  -3.55%        (p=0.000 n=10+10)
Update1_Cockroach-8           356µs ± 2%     343µs ± 1%  -3.49%         (p=0.000 n=10+9)
Delete1_Cockroach-8           390µs ± 2%     389µs ± 2%    ~           (p=0.529 n=10+10)
TrackChoices1_Cockroach-8     280µs ± 1%     269µs ± 1%  -3.61%        (p=0.000 n=10+10)
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7015)

<!-- Reviewable:end -->

petermattis committed:
storage: use a "distinct" batch in Replica.handleRaftReady()

A minor optimization that affects small transactions.

Fixes #6949.

name                       old time/op    new time/op    delta
Insert1_Cockroach-8           212µs ± 2%     204µs ± 1%  -3.55%        (p=0.000 n=10+10)
Update1_Cockroach-8           356µs ± 2%     343µs ± 1%  -3.49%         (p=0.000 n=10+9)
Delete1_Cockroach-8           390µs ± 2%     389µs ± 2%    ~           (p=0.529 n=10+10)
TrackChoices1_Cockroach-8     280µs ± 1%     269µs ± 1%  -3.61%        (p=0.000 n=10+10)

BramGruneir commented:
LGTM

tschottdorf commented on storage/replica.go:
```diff
@@ -1421,20 +1421,25 @@ func (r *Replica) handleRaftReady() error {
 		}
 		// TODO(bdarnell): update coalesced heartbeat mapping with snapshot info.
 	}
+	// We know that all of the writes from here forward will be to distinct keys.
```
Isn't the right thing to require here that none of the keys written have previously been mutated in the batch? Reads get passed through to the underlying engine (not the batch). We definitely mutate a lot of keys when we apply a snapshot, conceivably the hard state, last index, ... (maybe we don't currently, but putting this `Distinct()` business in for minor gains seems like a good recipe for a future two-day debugging session).

petermattis replied:
I was curious about this. Would it be more palatable to only enable this usage of a "distinct" batch if a snapshot isn't being applied?

tschottdorf replied:
Yes, I think so (though of course ideally we also want to use `Distinct()` in applySnapshot). I'm still somewhat uncomfortable with the semantics of `Distinct()` - they essentially require you to know about all prior writes to the batch. Remind me, why didn't we make `Distinct()` read from the underlying batch? It was because that opened up new questions about what would happen if the batch was written to while a Distinct batch was also active?
I'm thinking maybe the best way to go about it would be making the batch inactive as long as it has an open `Distinct()` batch (i.e. a write to the underlying batch would fail), and making `distinctBatch.Commit()` do nothing but flush to the underlying batch (which would then have to be committed)? The upshot would be that `Distinct()` wouldn't have to worry about what happened to the batch earlier, and that use of the original batch is disallowed until we're done with `Distinct()` mode. I think that's congruent with our usage and avoids the weird global perspective that's currently required for reasoning about when `Distinct()` applies.

petermattis replied:
My recollection is that `distinctBatch` has its current semantics so that you didn't have to recall when you created it to know what writes it would see. That is simple at one level, but as you've noted it adds cognitive complexity at another. Let me take a look at making the change you suggest. I don't think it will be difficult.

petermattis replied:
Well, this was mostly straightforward except for one large gotcha: the current usage of `Distinct()` within `Replica.{Put,ConditionalPut}` would now cause a flush on every request which completely destroys the distinct optimization.

petermattis replied:
Ok, got it figured out. I've added another commit which changes the semantics of `Distinct()`. PTAL.

petermattis committed:
storage/engine: change the semantics of "distinct" batches

A distinct batch will now read previous writes made to the parent batch,
but won't read its own writes. Also make it illegal to use the parent
batch while the distinct batch is open.

bdarnell commented on storage/replica.go:
```diff
@@ -1421,20 +1421,25 @@ func (r *Replica) handleRaftReady() error {
 	batch := r.store.Engine().NewBatch()
 	defer batch.Close()
 	if !raft.IsEmptySnap(rd.Snapshot) {
+		// TODO(peter): Could use a "distinct" batch within applySnapshot, but we
```
This comment points to the problems with this design: applySnapshot and not its caller should be responsible for deciding whether a distinct batch should be used. It's currently unsafe to use a distinct batch in applySnapshot, but I think #6878 would make it safe (but it may not be useful because applySnapshot will try to read its writes from the batch, so it would have to flush to c++). 

Another possibility: snapshot writes don't need to be atomic with the rest of this method. We use one batch for efficiency, bit it might be better to separate them.

petermattis replied:
Doesn't seem worth the effort to optimize this handling of snapshots as they should be much rarer than normal raft processing. Should I remove the TODO?

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. Do you see any chances to consolidate some of the interplay between `Blind` and `DistinctSpans`? They seem similar, but not quite the same.

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> LGTM
</details>


---

Reviewed 1 of 1 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1424 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-KJHAYc7kQMRNMfZ0tpr-r1-1424:-KJM9g7ijmQLMYJct4YZ:1230684567) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b812d957cbc54fddb7cede6d6114be5ea7fd39b/storage/replica.go#L1424)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Doesn't seem worth the effort to optimize this handling of snapshots as they should be much rarer than normal raft processing. Should I remove the TODO?
</details>
Using a new batch for snapshot doesn't seem like a bad idea conceptually, though there's no immediate upshot if we don't use `Distinct()`.
For the snapshot case, the go-side caching of mutations is a penalty because we write everything to Go and then force a flush down to C++, doubling the work, correct?
I don't think the TODO is worth it at this point, but maybe it could be converted into a comment that explains why there's not much to be gained.

Snapshots are not frequent (normally) but they are pretty large. Maybe optimizing them is still useful at some point.

---

*[storage/engine/engine.go, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-KJMWL6ZZ7PTAGF2d3Fy:-KJMWL6ZZ7PTAGF2d3Fz:-1825252371) ([raw file](https://github.com/cockroachdb/cockroach/blob/07c227d5919cebef6483a41f821839f3eadda0ca/storage/engine/engine.go#L92)):*
> ```Go
> // Reader is the read interface to an engine's data.
> type Reader interface {
> 	// Close closes the reader, freeing up any outstanding resources.
> ```

The comment here is somewhat misleading. For a normal batch, `Close()` aborts the batch. For a `Distinct()` batch, it frees the snapshot but keeps the writes in the original batch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-:-KJMWdPlZ768D8F_We7w:459823757)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> :lgtm:. Do you see any chances to consolidate some of the interplay between `Blind` and `DistinctSpans`? They seem similar, but not quite the same.
</details>


---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1424 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-KJHAYc7kQMRNMfZ0tpr-r1-1424:-KJNXEESqvydngeHZCjc:762903364) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b812d957cbc54fddb7cede6d6114be5ea7fd39b/storage/replica.go#L1424)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Using a new batch for snapshot doesn't seem like a bad idea conceptually, though there's no immediate upshot if we don't use `Distinct()`.
> For the snapshot case, the go-side caching of mutations is a penalty because we write everything to Go and then force a flush down to C++, doubling the work, correct?
> I don't think the TODO is worth it at this point, but maybe it could be converted into a comment that explains why there's not much to be gained.
> 
> Snapshots are not frequent (normally) but they are pretty large. Maybe optimizing them is still useful at some point.
</details>
Yeah, snapshots are large and (mostly) distinct, so there's more to gain by using a distinct batch with them, but they're rare enough that it's probably not worth the trouble and I'd remove the TODO. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-:-KJNZ-_goX0gV6GsT-27:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1424 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-KJHAYc7kQMRNMfZ0tpr-r1-1424:-KJNxMpSYmuvot-Oou6y:87244683) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b812d957cbc54fddb7cede6d6114be5ea7fd39b/storage/replica.go#L1424)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, snapshots are large and (mostly) distinct, so there's more to gain by using a distinct batch with them, but they're rare enough that it's probably not worth the trouble and I'd remove the TODO. 
</details>
Removed the TODO. 

Writing to Go and then flushing to C++ is not double the work of writing directly to C++ because the Go writes are much faster. I imagine we'll optimize applying snapshots at some point, though I suspect the initial optimization work will be on somehow streaming them instead of sending them in one request.

---

*[storage/engine/engine.go, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-KJMWL6ZZ7PTAGF2d3Fy:-KJNy-RFKP8gy8frCgPB:822029095) ([raw file](https://github.com/cockroachdb/cockroach/blob/07c227d5919cebef6483a41f821839f3eadda0ca/storage/engine/engine.go#L92)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The comment here is somewhat misleading. For a normal batch, `Close()` aborts the batch. For a `Distinct()` batch, it frees the snapshot but keeps the writes in the original batch.
</details>
Well, `Close()` does free up outstanding resources for the `Distinct()` batch.  Add some additional commentary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015)*
<!-- Sent from Reviewable.io -->

petermattis commented:
As you say, `Blind` and `DistinctSpans` are similar but not quite the same. `Blind` indicates that the key has never been written before (and thus there is nothing to read from RocksDB). `DistinctSpans` indicates that all of the keys in a batch or distinct and thus operations in the batch do not need to read their own writes. I don't see any way to combine the two. 

<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7015#-:-KJO3T4EWwI2dUWcAcuL:948932202)*
<!-- Sent from Reviewable.io -->
